node/src/components/storage.rs

```diff
@@ -1578,24 +1530,33 @@ fn initialize_block_body_v1_db(
 ) -> Result<(), LmdbExtError> {
     info!("initializing v1 block body database");
     let mut txn = env.begin_rw_txn()?;
+
+    let block_body_hash_to_header_map =
+        construct_block_body_to_block_header_reverse_lookup(&txn, block_header_db)?;
```

Comment:
How come we didn't need it before? Was the V1 DB initialized incorrectly?

Reply:
Also, won't it be slow? And get slower with each node restart? We're iterating over _all_ block headers in the database.

Reply:
We kind of _did_ need it - there was a bug here, actually :grimacing: 

It is possible for multiple blocks with the same deploy lists (for example: empty lists) and proposers to exist. These blocks would have the same block body hash, and so all of them would only get a single entry in the block body DB. Now, if we deleted one such block, its body hash would get marked for deletion and we would delete the body entry here - leaving all the other blocks without a corresponding entry in the block body DB.

The solution here, which I've used, is to only delete block bodies that aren't referred to by any header - but this means we have to collect all the bodies that _are_ referred to by some header, which is basically equivalent to constructing this index.